pipeline/src/pipeline.rs

```diff
@@ -807,132 +511,379 @@ impl<T: FieldElement> Pipeline<T> {
+        if self.artifact.parsed_asm_file.is_none() {
+            self.artifact.parsed_asm_file = Some(self.compute_parsed_asm_file()?);
         }
+
+        Ok(self.artifact.parsed_asm_file.as_ref().unwrap())
     }
 
-    pub fn analyzed_asm(mut self) -> Result<AnalysisASMFile<T>, Vec<String>> {
-        self.advance_to(Stage::AnalyzedAsm)?;
-        let Artifact::AnalyzedAsm(analyzed_asm) = self.artifact.unwrap() else {
-            panic!()
-        };
+    fn compute_resolved_module_tree(&mut self) -> Result<ASMProgram<T>, Vec<String>> {
+        let (path, parsed) = self.parsed_asm_file().unwrap().clone();
+
+        self.log("Loading dependencies and resolving references");
+        let resolved =
+            powdr_importer::load_dependencies_and_resolve(path, parsed).map_err(|e| vec![e])?;
+        self.diff_monitor.push(&resolved);
+
+        Ok(resolved)
+    }
+
+    pub fn resolved_module_tree(&mut self) -> Result<&ASMProgram<T>, Vec<String>> {
+        if self.artifact.resolved_module_tree.is_none() {
+            self.artifact.resolved_module_tree = Some(self.compute_resolved_module_tree()?);
+        }
+
+        Ok(self.artifact.resolved_module_tree.as_ref().unwrap())
+    }
+
+    fn compute_analyzed_asm(&mut self) -> Result<AnalysisASMFile<T>, Vec<String>> {
+        let resolved = self.resolved_module_tree().unwrap().clone();
+
+        self.log("Run analysis");
+        let analyzed_asm = powdr_analysis::analyze(resolved, &mut self.diff_monitor)?;
+        self.log("Analysis done");
+        log::trace!("{analyzed_asm}");
+
         Ok(analyzed_asm)
     }
 
-    pub fn analyzed_asm_ref(&mut self) -> Result<&AnalysisASMFile<T>, Vec<String>> {
-        self.advance_to(Stage::AnalyzedAsm)?;
-        match self.artifact.as_ref().unwrap() {
-            Artifact::AnalyzedAsm(analyzed_asm) => Ok(analyzed_asm),
-            _ => panic!(),
+    pub fn analyzed_asm(&mut self) -> Result<&AnalysisASMFile<T>, Vec<String>> {
+        if self.artifact.analyzed_asm.is_none() {
+            self.artifact.analyzed_asm = Some(self.compute_analyzed_asm()?);
         }
+
+        Ok(self.artifact.analyzed_asm.as_ref().unwrap())
+    }
+
+    fn compute_constrained_machine_collection(
+        &mut self,
+    ) -> Result<AnalysisASMFile<T>, Vec<String>> {
+        let analyzed_asm = self.analyzed_asm().unwrap().clone();
+
+        Ok(powdr_analysis::convert_vms_to_constrained(
+            analyzed_asm,
+            &mut self.diff_monitor,
+        ))
+    }
+
+    pub fn constrained_machine_collection(&mut self) -> Result<AnalysisASMFile<T>, Vec<String>> {
+        if self.artifact.constrained_machine_collection.is_none() {
+            self.artifact.constrained_machine_collection =
+                Some(self.compute_constrained_machine_collection()?);
+        }
+
+        Ok(self
+            .artifact
+            .constrained_machine_collection
+            .as_ref()
+            .unwrap()
+            .clone())
```

Comment:
Actually, if the reason that it takes the input by value is that it can move a lot of stuff from the input, I think it's fine. Otherwise, we'd just clone somewhere else.